Min integer Challenge

tamirwaleed opened the pull request:
---
name: solution review 
about: A template PR for code review with a checklist
---

<!--
  make this PR easy to find:

  - assign: yourself
  - link it to the corresponding challenge issue on the project board
-->

<!-- describe the challenge you will solve -->

## Behavior

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

tamirwaleed committed:
Added my learning goals, completed the min integer challenge

tamirwaleed committed:
Fixed formatting

tamirwaleed commented:
Completed the self-assigned Min Integer challenge, I've also added my learning goals to the markdown file in the collaboration directory

Alaa-Elgozouli reviewed (changes requested):
Hi Tamir, the code works well and all tests pass, however, I think there're few edits that need to be made before merging.

- The test class name should be in PascalCase. It should be "Class TestMinInteger(unittest.TestCase)".

- There's a small spelling mistake in line 32 in the test file. Should be "test_min_at_beginning".

- The code is fine and has no issues, however, it does more than its description. For example, the description states that the returns need to be an int, there're tests with floats and strings, so in this case the test should raise an error instead of passing, and for this to happen, there needs to be an assertion to make sure that all elements are int so it could return an int. I'd suggest adding 

`Assert isinstance(element, int), "all elements must be integers"`

and an assertion error if input is not int.

- In the test file, in line 63, the return is "Brennan" which in this case does not necessarily add up to how the function works. The string with the least amount of letters is "my" and "is", however, the return is "Brennan" because it returns the least index which in this case is 1, and not actually the word with the least amount of letters, and I don't think this aligns with the main purpose of the function.

- Since the function's name is "Min-Integer" I'd suggest making it only about integers without adding strings or floats. It would make sense to add floats incase both the return and elements statements state that they could be either int or float. However, I think it'd be better to exclude strings in this case. 

Overall, the code works just fine but it does more than what the function description says. Best of luck!!

Alaa-Elgozouli committed:
Merge branch 'main' into Min-Integer

tamirwaleed committed:
Fixed PascalCase & Spelling Error

tamirwaleed commented:
Hello Alaa, I greatly appreciate the attention to details that you've put into reviewing the code.  

Accordingly:
1.  I've fixed the PascalCase requirement, thus fulfilling the checklist completely.
2. I've also fixed the spelling mistake, great catch!

As for your other suggestions, I've put them into consideration, and I agree the function does more than required. Although, this was my interpretation of the behavior, I'll definitely consider this perspective in any future challenges going forward. Truly appreciated, and if there any other suggestions please let me know. Thank you!

Alaa-Elgozouli reviewed (approved):
All is good!!
